File: app/utils/mq/operations.js
const Joi = require('joi');

const Algorithm = require('../../algorithm');
const { getLogger } = require('../logger');
const { getMQ } = require('.');
const { imagingSchema } = require('../../schemas');

const logger = getLogger();

class MQOperations {
  constructor(inQueue, outExchange, options = {}) {
    this.channel = getMQ();
    this.algorithm = new Algorithm();

    this.inQueue = inQueue;
    this.outExchange = outExchange;

    this.PERSISTENT = true;
    this.FETCH_COUNT = 1;
    this.NO_ACK = false;
    this.REQUEUE_ON_ALG_ERR = false;
    this.REQUEUE_ON_PUB_ERR = true;

    this.options = options;
    this.options.persistent = this.PERSISTENT;
  }

  // eslint-disable-next-line class-methods-use-this
  _getRoutingKey(diagnosis) {
    return `${diagnosis.diagnosis}`;
  }

  async _publish(key, content) {
    // eslint-disable-next-line max-len
    await this.channel.publish(this.outExchange, key, Buffer.from(JSON.stringify(content)), this.persistent);
  }

  async _msgHandler(msg) {
    let imaging = {};
    let diagnosis = {};

    try {
      imaging = JSON.parse(msg.content.toString());
      Joi.assert(imaging, imagingSchema);
      logger.info(`Consumed imaging ${imaging._id}`);

      diagnosis = this.algorithm.run(imaging);
    }
    catch (err) {
      logger.error(err);
      this.channel.reject(msg, this.REQUEUE_ON_ALG_ERR);
      return logger.error(`Rejected imaging ${imaging._id} with requeue=${this.REQUEUE_ON_ALG_ERR}`);
    }

    try {
      const routingKey = this._getRoutingKey(diagnosis);
      await this._publish(routingKey, diagnosis);
      logger.info(`Published ${diagnosis.diagnosis} diagnosis of imaging ${imaging._id} to ${this.outExchange} exchange`);
    }
    catch (err) {
      logger.error(err);
      this.channel.reject(msg, this.REQUEUE_ON_PUB_ERR);
      return logger.error(`Rejected imaging ${imaging._id} with requeue=${this.REQUEUE_ON_PUB_ERR}`);
    }

    this.channel.ack(msg);
    return logger.info(`Acked imaging ${imaging._id}`);
  }

  async consume() {
    await this.channel.prefetch(this.FETCH_COUNT);

    await this.channel.consume(
      this.inQueue,
      this._msgHandler.bind(this),
      { noAck: this.NO_ACK },
    );
  }
}

module.exports = MQOperations;

File: app/utils/mq/operations.py
import json

import aiormq
from pconf import Pconf

from app.algorithm import Algorithm
from app.schemas import ImagingSchema
from app.utils.logger import get_logger
from app.utils.mq import get_mq

logger = get_logger()

class MQOperations:
    def __init__(self, in_queue, out_exchange):
        self.channel = get_mq()
        self.algorithm = Algorithm()

        self.in_queue = in_queue
        self.out_exchange = out_exchange

        self.DELIVERY_MODE = 2
        self.FETCH_COUNT = 1
        self.NO_ACK = False
        self.REQUEUE_ON_ALG_ERR = False
        self.REQUEUE_ON_PUB_ERR = True

    @staticmethod
    def _get_routing_key(diagnosis):
        return f"{diagnosis.get('diagnosis')}"

    async def _publish(self, key, content):
        await self.channel.basic_publish(
            exchange=self.out_exchange,
            routing_key=key,
            body=json.dumps(content).encode('utf-8'),
            properties=aiormq.spec.Basic.Properties(
                delivery_mode=self.DELIVERY_MODE,
            ),
        )

    async def _msg_handler(self, message):
        imaging = {}
        diagnosis = {}

        try:
            imaging = json.loads(message.body.decode('utf-8'))
            ImagingSchema().load(imaging)
            logger.info(f"Consumed imaging {imaging.get('_id')}")

            diagnosis = self.algorithm.run(imaging)
        except Exception as err: # pylint: disable=broad-except
            logger.error(err)
            await self.channel.basic_reject(delivery_tag=message.delivery.delivery_tag, requeue=self.REQUEUE_ON_ALG_ERR)
            
            dead_log = f" to {Pconf.get().get('AMQP_DEAD_EXCHANGE')} exchange" if self.REQUEUE_ON_ALG_ERR else ''
            return logger.error(f"Rejected imaging {imaging.get('_id')} with requeue={self.REQUEUE_ON_ALG_ERR}{dead_log}")

        try:
            routing_key = self._get_routing_key(diagnosis)
            await self._publish(routing_key, diagnosis)
            logger.info(f"Published {diagnosis.get('diagnosis')} diagnosis of imaging {imaging.get('_id')} to {self.out_exchange} exchange")
        except Exception as err: # pylint: disable=broad-except
            logger.error(err)
            await self.channel.basic_reject(delivery_tag=message.delivery.delivery_tag, requeue=self.REQUEUE_ON_PUB_ERR)
            
            dead_log = f" to {Pconf.get().get('AMQP_DEAD_EXCHANGE')} exchange" if self.REQUEUE_ON_PUB_ERR else ''
            return logger.error(f"Rejected imaging {imaging.get('_id')} with requeue={self.REQUEUE_ON_PUB_ERR}{dead_log}")

        await self.channel.basic_ack(delivery_tag=message.delivery.delivery_tag)
        return logger.info(f"Acked imaging {imaging.get('_id')}")

    async def consume(self):
        await self.channel.basic_qos(prefetch_count=self.FETCH_COUNT)
        await self.channel.basic_consume(queue=self.in_queue, consumer_callback=self._msg_handler)
